Add an optional SriovNetwork section: additional metaplugins

fedepaol opened the pull request:
In order to generate a network attachment definition that chains the SR-IOV CNI configuration with additional metaplugins (such as VRF, or firewall), we expose a new MetaPluginsConfig where to put the json list of the extra configurations, in the form of

```
{
	// meta plugin A config
},
{
	// meta plugin B config
}
```

fedepaol commented:
@zshi-redhat @pliurh please take a look.
If the approach sounds sane, I'll add e2e tests and docs.

zshi-redhat commented on bindata/manifests/cni-config/sriov-cni-config.yaml:
```diff
@@ -9,6 +9,10 @@ spec:
+{{- if .MetaPluginsConfigured -}}
+  "plugins": [
+    {
+  {{- end -}}  
```
nit: indent doesn't match with if?

fedepaol replied:
Done

zshi-redhat commented on bindata/manifests/cni-config/sriov-cni-config.yaml:
```diff
@@ -35,4 +39,10 @@ spec:
+}
+  {{- if .MetaPluginsConfigured -}}
+  ,
+  {{.MetaPlugins}}
```
Do we need curly braces for `{{.MetaPlugins}}`?

fedepaol replied:
I guess no:
The idea is to ask the users to fill the array, meaning the expectation is to have them to add the braces (and it's the only way to make it work if they want to chain more than one plugin):
```
{
	// meta plugin A config
},
{
	// meta plugin B config
}
```

zshi-redhat replied:
ack, this is different from how `ipam` is configured, we might need document the string format.

fedepaol replied:
Yes, wanted to make sure the approach looked fine first.
Will change this PR with the docs.

zshi-redhat reviewed (commented):
This PR may require rebase after https://github.com/openshift/sriov-network-operator/pull/372 is in.

fedepaol commented:
@zshi-redhat did a few changes, mostly driven by the failing test, such as a missing bracket, and the need for 0.4.0 version for the list.

fedepaol commented:
/retest

zshi-redhat commented:
looks good to me.

pliurh commented on api/v1/sriovnetwork_types.go:
```diff
@@ -42,6 +42,9 @@ type SriovNetworkSpec struct {
 	MaxTxRate *int `json:"maxTxRate,omitempty"`
+	// MetaPluginsConfig configuration to be used in order to chain metaplugins to the sriov interface returned
+	// by the operator.
+	MetaPluginsConfig string `json:"metaplugins,omitempty"`
```
Shall we use `runtime.RawExtension` to store the MetaPluginsConfig? It will be useful if we need to manipulate the content of the meta plugin config.

fedepaol replied:
Good point. I used string to be consistent with the ipam field, but I agree makes more sense to use RawExtension.
Will change

pliurh replied:
I want to change the `ipam` field to `runtime.RawExtension` as well. But I am not quite sure whether it will break the backward compatibility or not. We can address that in separate PR later.

fedepaol replied:
Question: if we go with the RawExtension, the user will need to put the yaml-ified version of the cni config, right?

For example 

```
metaplugins: |
    { "type": "firewall", "backend": "iptables" }
```

becomes something like

```
  metaplugins:
    plugins:
      - type: firewall
        backend: iptables

```

Am I missing something? If this is the case, I feel this is a bit worse for the user, as all the meta cni plugin examples are in the json form and they'll need to do the mental translation.
Thoughts?

pliurh replied:
Yeah. That is a problem. I agree it's better to keep using string here.

zshi-redhat replied:
yes, use string, or is rawMessage an option?

fedepaol replied:
I think rawMessage suffers of the same issue: we must declare the field as object, which eventually gets marshalled.
So in that case the user will have to convert the yaml to json.

If we want to leave the json format, I think `string` is the only option.

pliurh commented on bindata/manifests/cni-config/sriov-cni-config.yaml:
```diff
@@ -7,8 +7,16 @@ metadata:
 spec:
   config: '{
+{{- if .MetaPluginsConfigured -}}
+  "cniVersion":"0.4.0",
```
@zshi-redhat Does sriov-cni support cni spec 0.4.0? If yes, shall we always use version 0.4.0 here?

zshi-redhat replied:
I think the major change of 0.4.0 spec is the cmdCheck, we have a dummy interface interface for cmdCheck in sriov-cni.
maybe we shall wait until 0.4.0 is fully supported in sriov-cni.

fedepaol replied:
The 0.4.0 version is mandatory for using a list of plugins though.
@zshi-redhat is it fine to leave it as it is (setting only when a meta plugin is added)?

zshi-redhat replied:
Can we set cniVersion for meta plugin only?
Just checked sriov-cni, it accepts all CNI versions as supported, I think we shall be fine running sriov-cni with version 0.4.0 although I didn't test myself. 
Is 0.4.0 the only version supported by VRF? I see in the README page, it uses 0.3.1 as an example, but in the code, it has 0.4.0.

fedepaol commented:
@zshi-redhat @pliurh I reverted the CNI version to 0.3.1, changed the test to use the tuning plugin which is 0.3.1 compatible.

Please take another look

pliurh commented:
/lgtm
